src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +509,350 @@ pub fn process_participation_record_updates<
+}
+
+pub fn get_finality_delay<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> u64 {
+    get_previous_epoch(state, context) - state.finalized_checkpoint.epoch
+}
+
+pub fn is_in_inactivity_leak<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> bool {
+    get_finality_delay(state, context) > context.min_epochs_to_inactivity_penalty
+}
+
+pub fn get_eligible_validator_indices<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Vec<ValidatorIndex> {
+    let previous_epoch = get_previous_epoch(state, context);
+    state
+        .validators
+        .iter()
+        .enumerate()
+        .filter_map(|(i, v)| {
+            if is_active_validator(v, previous_epoch)
+                || (v.slashed && previous_epoch + 1 < v.withdrawable_epoch)
+            {
+                Some(i)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<ValidatorIndex>>()
```

Comment:
so one thing we can consider here is to avoid the allocation with the `Vec`....

perhaps it will ultimately not matter but i have been leaning towards using "existential" types for this with `impl Trait`:

```rust
fn get_eligible_validator_indices(...) -> impl Iterator<Item=ValidatorIndex> { ... }
```


and then you can drop the final `collect` you have here.

then the caller can decide what to do

Reply:
Having a little trouble getting the `impl Iterator` working -- trying to use lifetime values (like in `get_matching_target_attestations`), and I can't figure out how to align the return value. I'm getting an error:
```
expected enum `Option<&'a usize>`
found enum `Option<usize>`
```
...namely, idk how to get `usize` to be `&'a usize` 
https://gist.github.com/buchhlz2/9028b4387c10b30592edfefcb3d8e379

Reply:
i would just drop the ref... if you say `Item=ValidatorIndex` it will prolly work

i think the compiler is just doing a copy for you -- i don't really see a reason to work w/ ref's here

Reply:
Ah...I couldn't figure out how to use the lifetime unless there was a `&` involved....but yeah that works with `impl Iterator<Item = ValidatorIndex> + 'a` along with a `move` in the closure

Reply:
ah fair point...

i think what i am *trying* to do would actually imply a `cloned` after the `filter_map` and without making this a `move` closure...

if this is becoming a time sink, its fine to just `collect` into a `Vec` imo

Reply:
Oops, just saw this. Well, I opted for the `impl Iterator` and a closure `move` so lmk if I should refactor.